bindings/rust/src/bindings/mod.rs

```diff
@@ -42,6 +48,356 @@ pub struct KZGProof {
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct GenericBlob {
+    bytes: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub struct KzgSettings<
+    const FIELD_ELEMENTS_PER_BLOB: usize,
+    const BYTES_PER_BLOB: usize,
+    B: ValidatedBlob,
+> {
+    kzg_settings: KZGSettings,
+    _phantom: PhantomData<B>,
+}
+
+pub type MainnetKzgSettings =
+    KzgSettings<FIELD_ELEMENTS_PER_BLOB_MAINNET, BYTES_PER_BLOB_MAINNET, MainnetBlob>;
+pub type MinimalKzgSettings =
+    KzgSettings<FIELD_ELEMENTS_PER_BLOB_MINIMAL, BYTES_PER_BLOB_MINIMAL, MinimalBlob>;
+pub type GenericMainnetKzgSettings =
+    KzgSettings<FIELD_ELEMENTS_PER_BLOB_MAINNET, BYTES_PER_BLOB_MAINNET, GenericBlob>;
+pub type GenericMinimalKzgSettings =
+    KzgSettings<FIELD_ELEMENTS_PER_BLOB_MINIMAL, BYTES_PER_BLOB_MINIMAL, GenericBlob>;
+
+impl<const FIELD_ELEMENTS_PER_BLOB: usize, const BYTES_PER_BLOB: usize>
+    KzgSettings<FIELD_ELEMENTS_PER_BLOB, BYTES_PER_BLOB, GenericBlob>
+{
+    pub fn validate_blob(bytes: &[u8]) -> Result<GenericBlob, Error> {
+        let blob = slice_to_blob::<BYTES_PER_BLOB>(bytes)?;
+        Ok(GenericBlob {
+            bytes: blob.to_vec(),
+        })
+    }
+}
+
+pub trait ValidatedBlob: AsRef<[u8]> {}
+
+impl ValidatedBlob for MainnetBlob {}
+impl ValidatedBlob for MinimalBlob {}
+impl ValidatedBlob for GenericBlob {}
+
+pub trait KzgSettingsTrait {
+    type Blob;
+
+    fn new(kzg_settings: KZGSettings) -> Result<Self, Error>
+    where
+        Self: Sized;
+
+    fn load_trusted_setup_file(trusted_setup_file: &Path) -> Result<Self, Error>
+    where
+        Self: Sized;
+
+    fn load_trusted_setup(
+        g1_bytes: &[[u8; BYTES_PER_G1_POINT]],
+        g2_bytes: &[[u8; BYTES_PER_G2_POINT]],
+    ) -> Result<Self, Error>
+    where
+        Self: Sized;
+
+    fn blob_to_kzg_commitment(&self, blob: &Self::Blob) -> Result<KZGCommitment, Error>;
+
+    fn compute_kzg_proof(
+        &self,
+        blob: &Self::Blob,
+        z_bytes: &Bytes32,
+    ) -> Result<(KZGProof, Bytes32), Error>;
+
+    fn compute_blob_kzg_proof(
+        &self,
+        blob: &Self::Blob,
+        commitment_bytes: &Bytes48,
+    ) -> Result<KZGProof, Error>;
+
+    fn verify_kzg_proof(
+        &self,
+        commitment_bytes: &Bytes48,
+        z_bytes: &Bytes32,
+        y_bytes: &Bytes32,
+        proof_bytes: &Bytes48,
+    ) -> Result<bool, Error>;
+
+    fn verify_blob_kzg_proof(
+        &self,
+        blob: &Self::Blob,
+        commitment_bytes: &Bytes48,
+        proof_bytes: &Bytes48,
+    ) -> Result<bool, Error>;
+
+    fn verify_blob_kzg_proof_batch(
+        &self,
+        blobs: &[Self::Blob],
+        commitments_bytes: &[Bytes48],
+        proofs_bytes: &[Bytes48],
+    ) -> Result<bool, Error>;
+}
+
+impl<const FIELD_ELEMENTS_PER_BLOB: usize, const BYTES_PER_BLOB: usize, B: ValidatedBlob>
+    KzgSettingsTrait for KzgSettings<FIELD_ELEMENTS_PER_BLOB, BYTES_PER_BLOB, B>
+{
+    type Blob = B;
+
+    fn new(kzg_settings: KZGSettings) -> Result<Self, Error> {
```

Comment:
copying from telegram:
>>> given BYTES_PER_FIELD_ELEMENT is the same for either, it could just be one const generic KzgSettings<FIELD_ELEMENTS_PER_BLOB, B> instead right?
>>
>> Yep you can derive the bytes per blob from field elements per blob or vice versa
>
> yeah so can be either const for that
> probably BYTES_PER_BLOB since the Blob type uses that